CRM/Contribute/Form/Task/PDFLetterCommon.php

```diff
@@ -89,18 +89,19 @@ public static function postProcess(&$form, $formValues = NULL) {
         }
         $contact['is_sent'][$groupBy][$groupByID] = TRUE;
       }
-      // update dates (do it for each contribution including grouped recurring contribution)
-      //@todo - the 2 calls below bypass all hooks. Using the api would possibly be slower than one call but not than 2
+      // Update receipt/thankyou dates
+      $contributionParams = array('id' => $contributionId);
       if ($receipt_update) {
-        $result = CRM_Core_DAO::setFieldValue('CRM_Contribute_DAO_Contribution', $contributionId, 'receipt_date', $nowDate);
-        if ($result) {
-          $receipts++;
-        }
+        $contributionParams['receipt_date'] = $nowDate;
       }
       if ($thankyou_update) {
-        $result = CRM_Core_DAO::setFieldValue('CRM_Contribute_DAO_Contribution', $contributionId, 'thankyou_date', $nowDate);
-        if ($result) {
-          $thanks++;
+        $contributionParams['thankyou_date'] = $nowDate;
+      }
+      if ($receipt_update || $thankyou_update) {
+        $result = civicrm_api3('Contribution', 'create', $contributionParams);
+        if (empty($result['is_error'])) {
```

Comment:
This if is invalid - there will be an exception thrown rather than an error returned. There are no obvious reasons why an exception would be thrown so I think it's OK not to add any handling for the possibility.

Also - I'm not 100% sure about the ternarys below. I would make it
$receipts = ($receipt_update ? $receipts + 1 : $receipts);